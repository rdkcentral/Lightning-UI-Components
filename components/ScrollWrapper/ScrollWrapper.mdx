import { Canvas, Story } from '@storybook/addon-docs/blocks';
import { Theme } from '../../.storybook/utils';
import ScrollWrapper from '.';

# ScrollWrapper

A scrollable container for text content.

## Usage

`ScrollWrapper` requires three properties: `w`, `h`, and `content`:

```js
import { ScrollWrapper } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      ScrollWrapper: {
        type: ScrollWrapper,
        h: 450,
        w: 600,
        content: 'Text content long enough to scroll vertically'
      }
    };
  }
};
```

<Canvas>
  <Story id="scrollwrapper--basic" />
</Canvas>

The `h` and `w` properties will determine the viewable box for `ScrollWrapper`. When focused, `ScrollWrapper` will display a scroll bar to the right of the content and up/down scrolling will be handled.

The amount of scrolling per keypress is determined via the `scrollStep` property. Change it to allow smaller or larger scroll steps per keypress.

```js
this.tag('ScrollWrapper').scrollStep = 50;
```

### Styling

Below is the `styles` function for `ScrollWrapper`.

```js
const styles = theme => ({
  text: {
    ...theme.typography.body1,
    color: theme.palette.text.light.primary
  },
  scrollBar: {
    color: 0xffffffff,
    w: 16,
    h: 16
  },
  scrollBarContainer: {
    color: 0xaaffffff,
    w: 8
  }
});
```

To change the internal styles for `ScrollWrapper`, wrap it in the `withStyles` mixin.

**Example**

```js
import withStyles from '@lightning/ui/mixins/withStyles';
import BaseScrollWrapper from '@lightning/ui/components/ScrollWrapper';

const styles = theme => ({
  text: {
    ...theme.typography.body2
  }
})
class ScrollWrapper = withStyles(BaseScrollWrapper, styles);
```

## API

### Properties

name|type|required|default|description
--|--|--|--|--
content|string|true|-|text content to be displayed
h|number|true|0|height of viewable container
scrollStep|number|false|10|how many pixels to scroll by on every up/down keypress
w|number|true|0|width of viewable container

### Methods

#### resetScroll(): void

Resets the `y` value of both the content and the scroll bar.

### Events

#### $scrollChanged('endUp'|'endDown', this)

Event fired via `fireAncestors`, is triggered when scroll reaches the top or bottom of the scroll boundaries.

### Styles

Styled components can be adjusted by composing using [withStyles](?path=/docs/mixins-withstyles--basic). Below is the default `styles` object

<Theme theme={ScrollWrapper.styles} />