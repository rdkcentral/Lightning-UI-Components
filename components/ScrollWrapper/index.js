import lng from 'wpe-lightning';
import withStyles from '../../mixins/withStyles';

const styles = theme => ({
  text: {
    ...theme.typography.body1,
    color: theme.palette.text.light.primary
  },
  scrollBar: {
    color: 0xffffffff,
    w: 16,
    h: 16
  },
  scrollBarContainer: {
    color: 0xaaffffff,
    w: 8
  }
});

export default class ScrollWrapper extends withStyles(lng.Component, styles) {
  static _template() {
    return {
      clipping: true,
      ScrollContainer: {
        w: w => w,
        wordWrap: true
      },
      ScrollBarContainer: {
        alpha: 0,
        rect: true,
        w: this.styles.scrollBarContainer.w,
        x: w => w,
        color: this.styles.scrollBarContainer.color,
        ScrollBar: {
          rect: true,
          w: this.styles.scrollBar.w,
          h: this.styles.scrollBar.h,
          y: 0,
          color: this.styles.scrollBar.color
        }
      }
    };
  }

  _focus() {
    this._ScrollBarContainer.smooth = { alpha: 1 };
  }

  _unfocus() {
    this._ScrollBarContainer.smooth = { alpha: 0 };
  }

  _construct() {
    this._scrollStep = 10;
  }

  _firstEnable() {
    this._update();
  }

  _handleDown() {
    if (this._scrollContainerY + this._ScrollContainer.h > this.renderHeight) {
      const containerY = this._scrollContainerY - this._scrollStep;
      this._ScrollContainer.patch({
        smooth: {
          y: [
            containerY + this._ScrollContainer.h <= this.renderHeight
              ? this.renderHeight - this._ScrollContainer.h
              : containerY,
            {
              timingFunction: 'linear',
              duration: isNaN(this.scrollDuration) ? 0.2 : this.scrollDuration
            }
          ]
        }
      });
      const scrollBarY =
        this._scrollBarY + this._getScrollStep(this.renderHeight);
      this._ScrollBar.patch({
        smooth: {
          y: [
            scrollBarY < this.renderHeight - this.styles.scrollBar.h
              ? scrollBarY
              : this.renderHeight - this.styles.scrollBar.h,
            {
              timingFunction: 'linear',
              duration: isNaN(this.scrollDuration) ? 0.2 : this.scrollDuration
            }
          ]
        }
      });
      if (
        this._scrollContainerY + this._ScrollContainer.h <=
        this.renderHeight
      ) {
        this._autoScrollComplete = true;
        this.fireAncestors('$scrollChanged', 'endDown', this);
      }
    }
  }

  _handleUp() {
    if (this._scrollContainerY < 0) {
      const containerY = this._scrollContainerY + this._scrollStep;
      this._ScrollContainer.patch({
        smooth: {
          y: [
            containerY >= 0 ? 0 : this._scrollContainerY + this._scrollStep,
            {
              timingFunction: 'linear',
              duration: isNaN(this.scrollDuration) ? 0.2 : this.scrollDuration
            }
          ]
        }
      });
      const scrollBarY =
        this._scrollBarY - this._getScrollStep(this.renderHeight);
      this._ScrollBar.patch({
        smooth: {
          y: [
            scrollBarY > 0 ? scrollBarY : 0,
            {
              timingFunction: 'linear',
              duration: isNaN(this.scrollDuration) ? 0.2 : this.scrollDuration
            }
          ]
        }
      });

      if (this._scrollContainerY >= 0) {
        this.fireAncestors('$scrollChanged', 'endUp', this);
      }
    }
  }

  resetScroll() {
    this._ScrollContainer.patch({ y: 0 });
    this._ScrollBar.patch({ y: 0 });
    this._ScrollContainer.transition('y').finish();
    this._ScrollBar.transition('y').finish();
    delete this._ScrollContainer._transitions;
    delete this._ScrollBar._transitions;
    this._autoScrollComplete = false;
  }

  _update() {
    this._ScrollContainer.on('txLoaded', () => {
      const adjustW = this.renderWidth - this._ScrollBar.renderWidth;
      this._ScrollContainer.patch({
        y: 0,
        h: this._ScrollContainer.renderHeight,
        w: adjustW - 10
      });
      this._ScrollBarContainer.patch({
        h: this.renderHeight,
        x: adjustW,
        ScrollBar: {
          x: (this._ScrollBarContainer.w - this._ScrollBar.w) / 2
        }
      });
    });
    this._ScrollContainer.patch({
      h: 0,
      text: {
        ...this.styles.text,
        text: this._content
      }
    });
  }

  _getScrollStep(height) {
    const numSteps = Math.ceil(
      (this._ScrollContainer.h - this.renderHeight) / this._scrollStep
    );
    return (height - this.styles.scrollBar.h) / numSteps;
  }

  get autoScroll() {
    return this._autoScroll;
  }

  set autoScroll(val) {
    if (this._autoScroll !== val) {
      this._autoScroll = val;
    }
    this._setupAutoScroll();
  }

  _setupAutoScroll() {
    clearTimeout(this._startAutoScroll);
    clearTimeout(this._performAutoScrollTimer);

    if (this.autoScroll) {
      this._startAutoScroll = setTimeout(
        () => this._performAutoScroll(),
        isNaN(this.autoScrollDelay) ? 2000 : this.autoScrollDelay
      );
    }
  }

  get showScrollBar() {
    return this._showScrollBar;
  }

  set showScrollBar(val) {
    this._showScrollBar = val;
    this._ScrollBarContainer.visible = val;
  }

  _performAutoScroll() {
    if (this.autoScroll && !this._autoScrollComplete) {
      this._handleDown();
      this._performAutoScrollTimer = setTimeout(
        () => this._performAutoScroll(),
        isNaN(this.autoScrollSpeed) ? 200 : this.autoScrollSpeed
      );
    }
  }

  get content() {
    return this._content;
  }

  set content(content) {
    if (content !== this._content) {
      this._content = content;
      if (this.enabled) {
        this.resetScroll();
        this._setupAutoScroll();
        this._update();
      }
    }
  }

  get scrollStep() {
    return this._scrollStep;
  }

  set scrollStep(scrollStep) {
    if (scrollStep !== this._scrollStep) {
      this._scrollStep = scrollStep;
    }
  }

  get _scrollContainerY() {
    return this._ScrollContainer.transition('y').targetValue;
  }

  get _scrollBarY() {
    return this._ScrollBar.transition('y').targetValue;
  }

  get _ScrollContainer() {
    return this.tag('ScrollContainer');
  }

  get _ScrollBar() {
    return this.tag('ScrollBar');
  }

  get _ScrollBarContainer() {
    return this.tag('ScrollBarContainer');
  }
}
