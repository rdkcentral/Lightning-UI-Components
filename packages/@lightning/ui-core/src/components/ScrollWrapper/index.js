export { default as default } from './ScrollWrapper.js';
