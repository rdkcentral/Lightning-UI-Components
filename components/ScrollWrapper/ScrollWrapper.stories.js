import lng from '@lightningjs/core';

import ScrollWrapper from '.';
import mdx from './ScrollWrapper.mdx';

export default {
  title: 'ScrollWrapper',
  parameters: {
    docs: {
      page: mdx
    },
    argActions: {
      focused: (isFocused, component) => {
        component._getFocused = isFocused
          ? () => component.tag('ScrollWrapper')
          : () => {};
        component._refocus();
      }
    }
  },
  args: {
    autoScroll: false,
    autoScrollDelay: 2000,
    autoScrollSpeed: 200,
    focused: true,
    scrollDuration: 0.2,
    scrollStep: 10,
    showScrollBar: true
  },
  argTypes: {
    autoScroll: { control: 'boolean' },
    autoScrollDelay: { control: 'number' },
    autoScrollSpeed: { control: 'number' },
    focused: { control: 'boolean' },
    scrollChanged: { action: 'scrollChanged' },
    scrollDuration: { control: 'number' },
    scrollStep: { control: 'number' },
    showScrollBar: { control: 'boolean' }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        ScrollWrapper: {
          type: ScrollWrapper,
          h: 450,
          w: 600,
          scrollStep: args.scrollStep,
          autoScroll: args.autoScroll,
          scrollDuration: args.scrollDuration,
          showScrollBar: args.showScrollBar,
          autoScrollDelay: args.autoScrollDelay,
          autoScrollSpeed: args.autoScrollSpeed,
          content:
            'By activating, you agree that you want to enable cloud DVR technology to ' +
            'your Xfinity TV subscription on additional supported devices like computers and ' +
            'tablets, as well as the TV connected to your set-top DVR via Comcast’s network. If ' +
            'subscribed to X1 DVR with cloud technology, your acceptance also confirms that when you ' +
            'record programs on your set-top DVR you also want to save and store them automatically ' +
            'in Comcast’s network in several formats to (1) enable you to play them back (and ' +
            'navigate within them) on additional supported devices like computers and tablets, as ' +
            'well as the TV connected to your set-top DVR, and (2) optimize the video and audio ' +
            'quality of your viewing experience during playback.',
          signals: {
            scrollChanged: true
          }
        }
      };
    }

    $scrollChanged(type) {
      args.scrollChanged(type);
    }

    _getFocused() {
      if (args.focused) return this.tag('ScrollWrapper');
    }
  };

export const ContentArray = args =>
  class ContentArray extends lng.Component {
    static _template() {
      return {
        ScrollWrapper: {
          type: ScrollWrapper,
          h: 450,
          w: 600,
          wordWrap: true,
          scrollStep: args.scrollStep,
          autoScroll: args.autoScroll,
          scrollDuration: args.scrollDuration,
          showScrollBar: args.showScrollBar,
          autoScrollDelay: args.autoScrollDelay,
          autoScrollSpeed: args.autoScrollSpeed,
          content: [
            {
              text: 'Text Heading',
              style: {
                fontFace: 'XfinityBrownBold',
                alignContent: 'center'
              }
            },
            {
              text:
                'By activating, you agree that you want to enable cloud DVR technology to ' +
                'your Xfinity TV subscription on additional supported devices like computers and ' +
                'tablets, as well as the TV connected to your set-top DVR via Comcast’s network. If ' +
                'subscribed to X1 DVR with cloud technology, your acceptance also confirms that when you ' +
                'record programs on your set-top DVR you also want to save and store them automatically ' +
                'in Comcast’s network in several formats to (1) enable you to play them back (and ' +
                'navigate within them) on additional supported devices like computers and tablets, as ' +
                'well as the TV connected to your set-top DVR, and (2) optimize the video and audio ' +
                'quality of your viewing experience during playback.',
              style: {
                alignContent: 'left',
                fontSize: '30'
              }
            }
          ],
          signals: {
            scrollChanged: true
          }
        }
      };
    }

    $scrollChanged(type) {
      args.scrollChanged(type);
    }

    _getFocused() {
      if (args.focused) return this.tag('ScrollWrapper');
    }
  };
