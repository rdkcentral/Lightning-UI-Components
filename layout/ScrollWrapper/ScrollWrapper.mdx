import { Canvas, Story } from '@storybook/addon-docs/blocks';
import { Theme } from '../../.storybook/utils';
import ScrollWrapper from '.';

# ScrollWrapper

A scrollable container for text content.

## Usage

`ScrollWrapper` requires three properties: `w`, `h`, and `content`:

```js
import { ScrollWrapper } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      ScrollWrapper: {
        type: ScrollWrapper,
        h: 450,
        w: 600,
        content: 'Text content long enough to scroll vertically'
      }
    };
  }
};
```
<Canvas>
  <Story id="layout-scrollwrapper--basic" />
</Canvas>

Alternatively, an array of objects containing `text` and `style` keys may be provided as `content`. Styles provided here will override the ones set using the [`withStyles` mixin](#styling) for it's section of text:
```js
import { ScrollWrapper } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      ScrollWrapper: {
        type: ScrollWrapper,
        h: 450,
        w: 600,
        content: [
          {
            text: 'Text Heading',
            style: {
              fontFace: 'XfinityBrownBold',
              alignContent: 'center'
            }
          },
          {
            text: 'Text content long enough to scroll vertically'
          }
        ]
      }
    };
  }
};
```
<Canvas>
  <Story id="layout-scrollwrapper--content-array" />
</Canvas>

The `h` and `w` properties will determine the viewable box for `ScrollWrapper`. When focused, `ScrollWrapper` will display a scroll bar to the right of the content and up/down scrolling will be handled.

The amount of scrolling per keypress is determined via the `scrollStep` property. Change it to allow smaller or larger scroll steps per keypress.

Additionally, `ScrollWrapper` supports autoScroll flag, which will auto scroll the content. The delay to start is set via `autoScrollDelay` and
speed via `autoScrollSpeed` both in ms. You can also hide the scrollBar with `showScrollBar`

```js
this.tag('ScrollWrapper').scrollStep = 50;
```

### Styling

Below is the `styles` function for `ScrollWrapper`.

```js
const styles = theme => ({
  text: {
    ...theme.typography.body1,
    color: theme.palette.text.light.primary
  },
  scrollBar: {
    fillColor: theme.palette.grey[80],
    strokeColor: theme.palette.grey[10],
    stroke: 5,
    w: 32,
    h: 32,
    radius: 16
  },
  scrollBarContainer: {
    color: theme.palette.grey[50],
    progressColor: theme.palette.grey[10],
    w: 8,
    radius: 4
  }
});
```

To change the internal styles for `ScrollWrapper`, wrap it in the `withStyles` mixin.

**Example**

```js
import withStyles from '@lightning/ui/mixins/withStyles';
import BaseScrollWrapper from '@lightning/ui/components/ScrollWrapper';

const styles = theme => ({
  text: {
    ...theme.typography.body2
  }
})
class ScrollWrapper = withStyles(BaseScrollWrapper, styles);
```

## API

### Properties

name|type|required|default|description
--|--|--|--|--
content|string\|array|true|-|text content to be displayed. Can be passed as a string, or array of objects if varying text styles are needed
h|number|true|0|height of viewable container
autoScroll|boolean|false|false|auto scroll the content
autoScrollDelay|number|false|2000|delay before auto scroll starts
autoScrollSpeed|number|false|200|time delay before each scroll step
fadeContent|boolean|false|true|fade out content at the bottom of the ScrollWrapper
scrollDuration|number|false|0.2|animation duration for the scroll
scrollStep|number|false|10|how many pixels to scroll by on every up/down keypress
showScrollBar|boolean|false|true|show the scroll bar when focused
w|number|true|0|width of viewable container

### Methods

#### resetScroll(): void

Resets the `y` value of both the content and the scroll bar.

### Events

#### $scrollChanged('endUp'|'endDown', this)

Event fired via `fireAncestors`, is triggered when scroll reaches the top or bottom of the scroll boundaries.

### Styles

Styled components can be adjusted by composing using [withStyles](?path=/docs/mixins-withstyles--basic). Below is the default `styles` object

<Theme theme={ScrollWrapper.styles} />
