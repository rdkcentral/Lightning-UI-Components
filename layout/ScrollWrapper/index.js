import lng from '@lightningjs/core';
import Base from '../../Base';
import { withExtensions } from '../../mixins';
import * as styles from './ScrollWrapper.styles';
import TextBox from '../../elements/TextBox';
import ScrollSlider from './ScrollSlider';

class ScrollWrapper extends Base {
  static get __themeStyle() {
    return styles;
  }

  static _template() {
    return {
      clipping: true,
      FadeContainer: {
        ScrollContainer: {
          w: w => w,
          wordWrap: true
        }
      },
      Slider: {
        type: ScrollSlider,
        vertical: true,
        signals: {
          onChange: '_updateScrollContainerSize'
        }
      }
    };
  }

  static get __componentName() {
    return 'ScrollWrapper';
  }

  static get properties() {
    return [
      'autoScroll',
      'autoScrollDelay',
      'autoScrollSpeed',
      'content',
      'fadeContent',
      'scrollDuration',
      'scrollStep',
      'showScrollBar',
      'shouldWrap',
      'flexDirection'
    ];
  }

  static get tags() {
    return [
      'FadeContainer',
      'Slider',
      {
        name: 'ScrollContainer',
        path: 'FadeContainer.ScrollContainer'
      },
      {
        name: 'ScrollableText',
        path: 'ScrollContainer.ScrollableText'
      }
    ];
  }

  _construct() {
    super._construct();
    this._scrollStep = 10;
    this._shouldWrap = false;
    this._flexDirection = 'column';
    this._fadeContent = true;
    this._sliderWidth = 0;
  }

  $itemChanged() {
    // ScrollContainer uses flexbox, ensure a full stage layout so finalH is accurate
    this.stage.update();
    this._updateScrollWrapperLayout();
    this._updateAlpha();
  }

  _update() {
    if (this._contentChanged) {
      this._updateAutoScroll();
    }
    this._updateScrollContainer();
    this._updateScrollWrapperLayout();
    this._updateAlpha();
  }

  _updateAutoScroll() {
    this.resetScroll();
    this._setupAutoScroll();
    this._contentChanged = false;
  }

  _updateScrollWrapperLayout() {
    this._ScrollContainer.patch({
      w: this._contentWidth
    });
    this._updateFadeContainer();
    this._updateSlider();
  }

  _updateAlpha() {
    this._Slider.patch({
      smooth: {
        alpha: this.showScrollBar && this.mode === 'focused' ? 1 : 0
      }
    });
  }

  _updateFadeContainer() {
    const isScrollable = this._ScrollContainer.finalH > this.h;
    const shouldFade =
      this.fadeContent && isScrollable && !this._isEndContentVisible;

    this._FadeContainer.patch({
      h: this.h,
      w: this._contentWidth,
      y: this.style.contentMarginTop,
      x: this.style.contentMarginLeft,
      rtt: true,
      shader: shouldFade
        ? {
            type: lng.shaders.FadeOut,
            bottom: this.style.fadeHeight
          }
        : undefined
    });
  }

  _scrollContainerLoaded({ h }) {
    this._ScrollContainer.h = h;
    this._updateScrollWrapperLayout();
  }

  _updateScrollContainer() {
    if (!this.content) return;

    if (typeof this.content === 'string') {
      this._ScrollContainer.patch({
        flex: { direction: 'column' },
        ScrollableText: {
          h: 0,
          w: this._contentWidth,
          type: TextBox,
          content: this.content,
          wordWrap: true,
          wordWrapWidth: this._contentWidth,
          textStyle: this.style.text,
          signals: {
            textBoxChanged: '_scrollContainerLoaded'
          }
        }
      });
    } else if (Array.isArray(this.content)) {
      const content = {};
      this.content.forEach((item, index) => {
        const id = `ScrollText${index}`;
        content[id] = {
          w: this._contentWidth,
          ...item
        };
        if (item.text) {
          content[id] = {
            type: TextBox,
            content: item.text,
            wordWrap: true,
            wordWrapWidth: this._contentWidth,
            textStyle: {
              ...this.style.text,
              ...item.style
            }
          };
        }
      });
      this._ScrollContainer.patch({
        flex: { direction: this.flexDirection, wrap: this.shouldWrap },
        ...content
      });
    }
  }

  _scrollDown() {
    const hasContentToScrollTo =
      this._scrollContainerY + this._ScrollContainer.finalH > this.renderHeight;
    if (hasContentToScrollTo) {
      const targetY = this._scrollContainerY - this.scrollStep;
      const canScrollByStepLength =
        targetY + this._ScrollContainer.finalH > this.renderHeight;
      const scrollEndY =
        this.renderHeight -
        this._ScrollContainer.finalH -
        this.style.contentMarginTop;

      this._ScrollContainer.patch({
        smooth: {
          y: [
            canScrollByStepLength ? targetY : scrollEndY,
            this._scrollAnimation
          ]
        }
      });

      const hasScrolledToEnd =
        this._scrollContainerY + this._ScrollContainer.finalH <= this.h;

      if (hasScrolledToEnd) {
        this._isEndContentVisible = true;
        this._autoScrollComplete = true;
        this.fireAncestors('$scrollChanged', 'endDown', this);
        this._updateFadeContainer();
      }
    }
  }

  _scrollUp() {
    const canScrollUp = this._scrollContainerY < 0;

    if (canScrollUp) {
      const targetY = this._scrollContainerY + this.scrollStep;
      const canScrollByStepLength = targetY < 0;

      this._ScrollContainer.patch({
        smooth: {
          y: [canScrollByStepLength ? targetY : 0, this._scrollAnimation]
        }
      });

      const isScrollable =
        this._scrollContainerY + this._ScrollContainer.finalH >
        this.renderHeight;
      if (isScrollable) {
        this._autoScrollComplete = false;
      }

      if (this._scrollContainerY >= 0) {
        this.fireAncestors('$scrollChanged', 'endUp', this);
      }

      if (this._isEndContentVisible) {
        this._isEndContentVisible = false;
        this._updateFadeContainer();
      }
    }
  }

  resetScroll() {
    this._ScrollContainer.patch({ y: 0 });
    this._Slider.patch({ value: 0 });
    this._ScrollContainer.transition('y').finish();
    delete this._ScrollContainer._transitions;
    this._autoScrollComplete = false;
  }

  _setAutoScroll(val) {
    if (this._autoScroll !== val) {
      this._autoScroll = val;
    }
    this._setupAutoScroll();
    return val;
  }

  _setupAutoScroll() {
    clearTimeout(this._startAutoScroll);
    clearTimeout(this._performAutoScrollTimer);

    if (this.autoScroll) {
      this._startAutoScroll = setTimeout(
        () => this._performAutoScroll(),
        isNaN(this.autoScrollDelay) ? 2000 : this.autoScrollDelay
      );
    }
  }

  _setShowScrollBar(val) {
    this._Slider.smooth = { alpha: val ? 1 : 0 };
    return val;
  }

  _performAutoScroll() {
    if (this.autoScroll && !this._autoScrollComplete) {
      this._Slider._handleDown();
      this._performAutoScrollTimer = setTimeout(
        () => this._performAutoScroll(),
        isNaN(this.autoScrollSpeed) ? 200 : this.autoScrollSpeed
      );
    }
  }

  _setContent(content) {
    if (content !== this._content && this.enabled) {
      this._contentChanged = true;
    }
    return content;
  }

  _updateScrollContainerSize(sliderValue, slider) {
    if (this._sliderWidth !== slider._Container.h) {
      this._sliderWidth = slider._Container.h;
      this._updateScrollContainer();
    }
  }

  _updateSlider() {
    // height of off screen items that can be scrolled to
    const scrollHeight = this._ScrollContainer.finalH - this.renderHeight;
    // number of steps to scroll to bottom of ScrollContainer
    const contentScrollSteps = Math.ceil(scrollHeight / this.scrollStep);
    // max value of slider
    const sliderMax = this.renderHeight;
    // distance slider should move on each key press
    const sliderStep = sliderMax / contentScrollSteps;

    // This is a vertical slider, so w is actually controlling the height
    this._Slider.patch({
      palette: this.palette,
      x: this.w - this._sliderWidth,
      w: sliderMax,
      min: 0,
      max: sliderMax,
      step: sliderStep,
      onUp: this._scrollUp.bind(this),
      onDown: this._scrollDown.bind(this)
    });
  }

  get _contentWidth() {
    return (
      this.w -
      this.style.contentMarginLeft -
      this.style.sliderMarginLeft -
      this._sliderWidth
    );
  }

  get _scrollContainerY() {
    return this._ScrollContainer.transition('y').targetValue;
  }

  get _scrollAnimation() {
    const duration = isNaN(this.scrollDuration)
      ? this.style.scroll.duration
      : this.scrollDuration;
    return {
      ...this.style.scroll,
      duration
    };
  }

  _getFocused() {
    return this._Slider;
  }
}

export default withExtensions(ScrollWrapper);
