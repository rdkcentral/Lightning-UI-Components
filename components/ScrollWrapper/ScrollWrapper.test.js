import TestUtils from '../lightning-test-utils';
import ScrollWrapper from '.';

const createScrollWrapper = TestUtils.makeCreateComponent(ScrollWrapper);

describe('ScrollWrapper', () => {
  let scrollWrapper, testRenderer;

  beforeEach(() => {
    [scrollWrapper, testRenderer] = createScrollWrapper({
      h: 100,
      w: 100,
      content:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin eu aliquam libero. Sed ipsum ligula, egestas et sollicitudin eget, pulvinar et neque. Curabitur commodo nisi sit amet ligula ultrices, a sodales ante consequat. Ut ullamcorper odio et erat sagittis volutpat. Cras consequat dolor in nisi sagittis, quis volutpat mi tempor. Praesent condimentum quis purus eget sodales. Praesent tempus suscipit felis, quis gravida massa tempor ut.'
    });
    testRenderer.update();
  });

  afterEach(() => {
    scrollWrapper = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
    expect(scrollWrapper._ScrollContainer.text.text).toBe(
      scrollWrapper.content
    );
    expect(scrollWrapper._ScrollContainer.h).toBeGreaterThan(scrollWrapper.h);
  });

  it('hides scroll bar on unfocus', () => {
    testRenderer.unfocus();
    testRenderer.update();
    expect(scrollWrapper._ScrollBarContainer.alpha).toBe(0);
  });

  it('scrolls up and down', () => {
    expect(scrollWrapper._scrollContainerY).toBe(0);
    expect(scrollWrapper._scrollBarY).toBe(0);

    testRenderer.keyPress('Down');

    expect(scrollWrapper._scrollContainerY).toBeLessThan(0);
    expect(scrollWrapper._scrollBarY).toBeGreaterThan(0);

    testRenderer.keyPress('Up');

    expect(scrollWrapper._scrollContainerY).toBe(0);
    expect(scrollWrapper._scrollBarY).toBe(0);
  });

  it('does not scroll beyond boundaries', () => {
    testRenderer.keyPress('Up');
    expect(scrollWrapper._scrollContainerY).toBe(0);
    expect(scrollWrapper._scrollBarY).toBe(0);

    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');

    expect(
      scrollWrapper._scrollContainerY + scrollWrapper._ScrollContainer.h
    ).toBe(scrollWrapper.h);

    testRenderer.keyPress('Down');
    expect(
      scrollWrapper._scrollContainerY + scrollWrapper._ScrollContainer.h
    ).toBe(scrollWrapper.h);
  });

  it('resets scroll', () => {
    testRenderer.keyPress('Down');
    testRenderer.keyPress('Down');
    scrollWrapper.resetScroll();
    testRenderer.update();
    expect(scrollWrapper._ScrollContainer.y).toBe(0);
    expect(scrollWrapper._ScrollBar.y).toBe(0);
  });

  it('gets and sets scrollStep', () => {
    expect(scrollWrapper.scrollStep).toBe(10);
    scrollWrapper.scrollStep = 5;
    expect(scrollWrapper._scrollStep).toBe(5);
  });
});
