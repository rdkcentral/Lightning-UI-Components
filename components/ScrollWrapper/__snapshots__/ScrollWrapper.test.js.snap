// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollWrapper renders 1`] = `
Object {
  "Component": Object {
    "ScrollBarContainer": Object {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "ScrollBar": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 16,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "ScrollBar",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "ScrollBar",
          ],
          "texture": Object {
            "type": "RectangleTexture",
          },
          "visible": true,
          "w": 16,
          "x": -4,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": "aaffffff",
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 100,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "ScrollBarContainer",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "ScrollBarContainer",
      ],
      "texture": Object {
        "type": "RectangleTexture",
      },
      "visible": true,
      "w": 8,
      "x": 84,
      "y": 0,
      "zIndex": 0,
    },
    "ScrollContainer": Object {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 214,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "ScrollContainer",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "ScrollContainer",
      ],
      "texture": Object {
        "fontFace": "XfinityStandardMedium",
        "fontSize": 28,
        "lineHeight": 40,
        "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin eu aliquam libero. Sed ipsum ligula, egestas et sollicitudin eget, pulvinar et neque. Curabitur commodo nisi sit amet ligula ultrices, a sodales ante consequat. Ut ullamcorper odio et erat sagittis volutpat. Cras consequat dolor in nisi sagittis, quis volutpat mi tempor. Praesent condimentum quis purus eget sodales. Praesent tempus suscipit felis, quis gravida massa tempor ut.",
        "type": "TextTexture",
        "verticalAlign": "middle",
        "w": 100,
      },
      "visible": true,
      "w": 74,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
